src/fides/api/util/consent_util.py

```diff
@@ -228,3 +247,281 @@ def get_or_create_fides_user_device_id_provided_identity(
+    db record that will be updated.
+
+    For PrivacyNoticeTemplates, returns a list of tuples with the data to be upserted, alongside
+    the existing PrivacyNoticeTemplate if it exists or None otherwise.
+
+    :param privacy_notice_updates: List of privacy notice schemas with ids: appropriate
+    for editing PrivacyNotices or upserting PrivacyNoticeTemplates
+    :param db: Session
+    :param model: one of PrivacyNotice or PrivacyNoticeTemplate
+    :return:
+    """
+    allow_create: bool = model == PrivacyNoticeTemplate
+    ignore_disabled: bool = model != PrivacyNoticeTemplate
+
+    # first we populate a map of privacy notices or templates in the db, indexed by ID
+    existing_notices: Dict[str, Union[PrivacyNotice, PrivacyNoticeTemplate]] = {}
+    for existing_notice in model.query(db).all():
+        existing_notices[existing_notice.id] = existing_notice
+
+    # then associate existing notices/templates with their creates/updates
+    # we'll return this set of data to actually process updates
+    updates_and_existing: List[
+        Tuple[
+            PrivacyNoticeWithId, Optional[Union[PrivacyNotice, PrivacyNoticeTemplate]]
+        ]
+    ] = []
+    for update_data in privacy_notice_updates:
+        update_data = transform_fields(
+            transformation=escape,
+            model=update_data,
+            fields=PRIVACY_NOTICE_ESCAPE_FIELDS,
+        )  # type: ignore
+        if update_data.id not in existing_notices:
+            if not allow_create:
+                raise HTTPException(
+                    status_code=HTTP_404_NOT_FOUND,
+                    detail=f"No {model.__class__.__name__} found for id {update_data.id}.",
+                )
+            updates_and_existing.append((update_data, None))
+        else:
+            updates_and_existing.append((update_data, existing_notices[update_data.id]))
+
+    # we temporarily store proposed update data in-memory for validation purposes only
+    validation_updates = []
+    for update_data, existing_notice in updates_and_existing:
+        # add the patched update to our temporary updates for validation
+        if existing_notice:
+            dry_update = existing_notice.dry_update(
+                data=update_data.dict(exclude_unset=True)
+            )
+            validate_privacy_notice_dry_update(
+                dry_update
+            )  # Checks consent mechanism + delivery location
+            validation_updates.append(
+                existing_notice.dry_update(data=update_data.dict(exclude_unset=True))
+            )
+            # and don't include it anymore in the existing notices used for validation
+            existing_notices.pop(existing_notice.id, None)
+        else:
+            # Add the pending created record to the the temporary updates for validation
+            validation_updates.append(model(**update_data.dict(exclude_unset=True)))
```

Comment:
New - for templates, we're adding the pending created record here if it didn't already exist

Reply:
this is nice, pretty straightforward branching. i remember all of this validation logic is so complex, so i agree with you that slotting this in here pays dividends, even if there's some branching. the core validation flow stays in tact, and that's better to have consolidated in one place rather than being duplicated.

nice job thinking thru this, i'm sure it took a bit of mental grit to get this new use case integrated!